mycpp/gc_dict.h

```diff
@@ -307,49 +284,117 @@ void Dict<K, V>::clear() {
   len_ = 0;
 }
 
-// Returns the position in the array.  Used by dict_contains(), index(),
-// get(), and set().
-//
-// For now this does a linear search.
 // TODO:
-// - hash functions, and linear probing.
-// - resizing based on load factor
-//   - which requires rehashing (re-insert all items)
 // - Special case to intern Str* when it's hashed?  How?
 //   - Should we have wrappers like:
 //   - V GetAndIntern<V>(D, &string_key)
 //   - SetAndIntern<V>(D, &string_key, value)
 //   This will enable duplicate copies of the string to be garbage collected
+//
+// NOTE: This draws a lot of inspiration from this proposal for CPython's dict.
+// The links below are useful references.
+// https://code.activestate.com/recipes/578375/
+// https://mail.python.org/pipermail/python-dev/2012-December/123028.html
 template <typename K, typename V>
-int Dict<K, V>::position_of_key(K key) {
-  for (int i = 0; i < capacity_; ++i) {
-    int special = entry_->items_[i];  // NOT an index now
-    if (special == kDeletedEntry) {
-      continue;  // keep searching
+int Dict<K, V>::hash_and_probe(K key) const {
+  if (capacity_ == 0) {
+    return kNotFound;
+  }
+
+  // Hash the key onto a slot in the index. If the first slot is occupied, probe
+  // until an empty one is found.
+  unsigned h = hash_key(key);
+  int init_bucket = h % index_len_;
+
+  // If we see a tombstone along the probing path, stash it.
+  int open_slot = -1;
+
+  for (int i = 0; i < index_len_; ++i) {
+    // Start at init_bucket and wrap araound
+    // NOTE: if division becomes a hotspot, split this into two loops.
+    int slot = (i + init_bucket) % index_len_;
+
+    int kv_index = index_->items_[slot];
+    DCHECK(kv_index < len_);
+    // Optimistically this is the common case once the table has been populated.
+    if (kv_index >= 0) {
+      unsigned h2 = hash_key(keys_->items_[kv_index]);
+      if (h == h2 && keys_equal(keys_->items_[kv_index], key)) {
+        return slot;
+      }
+    }
+
+    if (kv_index == kEmptyEntry) {
+      if (open_slot != -1) {
+        slot = open_slot;
+      }
+      // If there isn't room in the entry arrays, tell the caller to resize.
+      return len_ < capacity_ ? slot : kNotFound;
     }
-    if (special == kEmptyEntry) {
-      return -1;  // not found
+
+    // Tombstone or collided keys unequal. Keep scanning.
+    DCHECK(kv_index >= 0 || kv_index == kDeletedEntry);
+    if (kv_index == kDeletedEntry && open_slot == -1) {
+      // NOTE: We only record the open slot here. We DON'T return it. If we're
+      // looking for a key that was writen before this tombstone was written to
+      // the index we should continue probing until we get to that key. If we
+      // get to an empty index slot or the end of the index then we know we are
+      // dealing with a new key and can safely replace the tombstone without
+      // disrupting any existing keys.
+      open_slot = slot;
+    }
+  }
+
+  if (open_slot != -1) {
```

Comment:
likewise it feels like when open_slot is set to an index that contains `kDeletedEntry`, then `len_ < capacity_` ?

So this could be

```
// We iterated through all the buckets, and did not find kEmptEntry

if (open_slot == -1) {  // We also did not find kDeletedEntry, so there's no place to put it
  return kNotFound;
} else {
  DCHECK(len_ < capacity_);
  return open_slot;  // Put it in the spot of a kDeletedEntry
}
```

?